include/pybind11/detail/type_caster_odr_guard.h

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2022 The Pybind Development Team.
+// All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+#pragma once
+
+#include "descr.h"
+
+#ifdef PYBIND11_TYPE_CASTER_ODR_GUARD_ON
+
+#    if !defined(PYBIND11_CPP20) && defined(__GNUC__) && !defined(__clang__)
+#        pragma GCC diagnostic ignored "-Wsubobject-linkage"
+#    endif
+
+#    include "../pytypes.h"
+#    include "common.h"
+#    include "typeid.h"
+
+#    include <cstdio>
+#    include <cstring>
+#    include <string>
+#    include <system_error>
+#    include <typeindex>
+#    include <typeinfo>
+#    include <unordered_map>
+#    include <utility>
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+inline std::unordered_map<std::type_index, std::string> &type_caster_odr_guard_registry() {
+    static std::unordered_map<std::type_index, std::string> reg;
```

Comment:
This could use the no-destructor idiom:

static std::unordered_map<std::type_index, std::string>& reg = 
    *new std::unordered_map<std::type_index, std::string>();

Reply:
Done (696b80a), thanks Laramie!

While I was at it, I introduced `using type_caster_odr_guard_registry_type = ...` to make the code nicer to read.